### PR TITLE
[Doc] Update our documents from ti.var to ti.field

### PR DESCRIPTION
**Pull request by Rullec**
Related issue = #1500

In this PR, documents are mainly updated  for `ti.var`->`ti.field`

TO-DO
- [ ] Test all code examples of our doc, after the new APIs are been implemented well.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Comment by Rullec** [comment]
@archibate thanks, good night!

File: docs/vector.rst [hunk 1]
```diff
@@ -212,28 +212,28 @@ Methods
     ::
 
         # Taichi-scope
-        a = ti.Vector([1.6, 2.3])
+        a = ti.Vector.field([1.6, 2.3])
         a.cast(ti.i32) # [2, 3]
 
 .. note::
-    Vectors are special matrices with only 1 column. In fact, ``ti.Vector`` is just an alias of ``ti.Matrix``.
+    Vectors are special matrices with only 1 column. In fact, ``ti.Vector.field`` is just an alias of ``ti.Matrix.field``.
```
Comment (archibate) [inline docs/vector.rst:219]:
```suggestion
    Vectors are special matrices with only 1 column. In fact, ``ti.Vector`` is just an alias of ``ti.Matrix``.
```

File: docs/vector.rst [hunk 2]
```diff
@@ -212,28 +212,28 @@ Methods
     ::
 
         # Taichi-scope
-        a = ti.Vector([1.6, 2.3])
+        a = ti.Vector.field([1.6, 2.3])
```
Comment (archibate) [inline docs/vector.rst:215]: Please don't rely too much on `s/.../.../g`, this is **allocating a vector as local variable**. Please revert all changes about `ti.Vector.field`.

File: docs/gui.rst [hunk 3]
```diff
@@ -53,7 +53,7 @@ Paint on a window
 .. function:: gui.set_image(img)
 
     :parameter gui: (GUI) the window object
-    :parameter img: (np.array or Tensor) tensor containing the image, see notes below
+    :parameter img: (np.array or field) field containing the image, see notes below
```
Comment (archibate) [inline docs/gui.rst:56]:
```suggestion
    :parameter img: (np.array or Field) field containing the image, see notes below
```

File: docs/vector.rst [hunk 4]
```diff
@@ -46,19 +46,19 @@ As a temporary local variable
...
 
-As global tensors of vectors
+As global fields of vectors
 ++++++++++++++++++++++++++++
 .. attribute:: a[p, q, ...][i]
 
-    :parameter a: (tensor of Vector) the vector
-    :parameter p: (scalar) index of the first tensor dimension
-    :parameter q: (scalar) index of the second tensor dimension
+    :parameter a: (field of Vector) the vector
```
Comment (archibate) [inline docs/vector.rst:59]:
```suggestion
    :parameter a: (Vector field) the vector
```
`vector field` sounds more intuitive.

File: docs/vector.rst (hunk 2 above)
Comment (Rullec) [reply docs/vector.rst:215]: Yes, it is so dangerous and I forget this local-global difference.
A revert will be made tomorrow, and I will check these changes manually one by one.

**Comment by Rullec** [comment]
> Thank for the hard work! But as you see **changing all tensor -> field** is a large changeset therefore very hard to review and causing the non-trivial problems listed above, so I'd suggest:
> 
> 1. **Only** replace `ti.var` -> `ti.field` in this PR.
> 2. Replace `tensor` -> `field` **in another PR**.
> 3. Replace `ti.Vector` -> `ti.Vector.field` in another PR.
>    So that I can review your PR faster, WDYT?

Very practical! 

Also under your reminder can I realize the difference that `ti.Vector` should only be used locally, and vice versa `ti.Vector.field`.Thanks!

Let's keep this PR for `ti.var`->`ti.field`.

**Comment by Rullec** [comment]
Update: `ti.var`->`ti.field` maybe mostly finished. 

I try to cover all cases though, there can still be some hidden words which I cannot get at once.

File: docs/compilation.rst [hunk 5]
```diff
@@ -26,12 +26,12 @@ Let's consider the following simple kernel:
       tensor[i] += delta
 
 
-We allocate two 1D tensors to simplify discussion:
+We allocate two 1D fields to simplify discussion:
```
Comment (archibate) [inline docs/compilation.rst:29]:
```suggestion
We allocate two 1D tensors to simplify discussion:
```

File: docs/internal.rst [hunk 6]
```diff
@@ -156,17 +156,17 @@ correspond to?
 Each ``SNode`` can have a different virtual-to-physical mapping. ``physical_index_position[i] == -1``
 means the ``i``-th virtual index does not corrspond to any physical index in this ``SNode``.
 
-``SNode`` s in handy dense tensors (i.e., ``a = ti.var(ti.i32, shape=(128, 256, 512))``)
+``SNode`` s in handy dense fields (i.e., ``a = ti.field(ti.i32, shape=(128, 256, 512))``)
```
Comment (archibate) [inline docs/internal.rst:159]:
```suggestion
``SNode`` s in handy dense tensors (i.e., ``a = ti.field(ti.i32, shape=(128, 256, 512))``)
```
Do not do `tensors` -> `fields` yet. We **only do `ti.var` -> `ti.field` in this PR**.

File: docs/layout.rst [hunk 7]
```diff
@@ -3,10 +3,10 @@
 Advanced dense layouts
 ======================
 
-Tensors (:ref:`scalar_tensor`) can be *placed* in a specific shape and *layout*.
+Fields (:ref:`scalar_field`) can be *placed* in a specific shape and *layout*.
```
Comment (archibate) [inline docs/layout.rst:6]:
```suggestion
Tensors (:ref:`scalar_tensor`) can be *placed* in a specific shape and *layout*.
```
Could you review yourself first?

File: docs/layout.rst [hunk 8]
```diff
@@ -3,10 +3,10 @@
...
 Defining a proper layout can be critical to performance, especially for memory-bound applications. A carefully designed data layout can significantly improve cache/TLB-hit rates and cacheline utilization. Although when performance is not the first priority, you probably don't have to worry about it.
 
-In Taichi, the layout is defined in a recursive manner. See :ref:`snode` for more details about how this works. We suggest starting with the default layout specification (simply by specifying ``shape`` when creating tensors using ``ti.var/Vector/Matrix``),
+In Taichi, the layout is defined in a recursive manner. See :ref:`snode` for more details about how this works. We suggest starting with the default layout specification (simply by specifying ``shape`` when creating fields using ``ti.field/Vector/Matrix``),
```
Comment (archibate) [inline docs/layout.rst:9]:
```suggestion
In Taichi, the layout is defined in a recursive manner. See :ref:`snode` for more details about how this works. We suggest starting with the default layout specification (simply by specifying ``shape`` when creating tensors using ``ti.field/Vector/Matrix``),
```

File: docs/matrix.rst [hunk 9]
```diff
@@ -38,11 +38,11 @@ As global tensors of matrices
     ::
 
         # Python-scope
-        a = ti.Matrix.var(3, 3, dt=ti.f32, shape=(5, 4))
+        a = ti.Matrix(3, 3, dt=ti.f32, shape=(5, 4))
 
 .. note::
 
-    In Python-scope, ``ti.var`` declares :ref:`scalar_tensor`, while ``ti.Matrix`` declares tensors of matrices.
+    In Python-scope, ``ti.field`` declares :ref:`scalar_field`, while ``ti.Matrix`` declares tensors of matrices.
```
Comment (archibate) [inline docs/matrix.rst:45]:
```suggestion
    In Python-scope, ``ti.field`` declares :ref:`scalar_tensor`, while ``ti.Matrix`` declares tensors of matrices.
```

File: docs/scalar_field.rst [hunk 10]
```diff
@@ -0,0 +1,174 @@
+.. _scalar_field:
```
Comment (archibate) [inline docs/scalar_field.rst:1]: Please revert the filename, that make my review hard, I can't view in diff mode if you insist to do so.

File: docs/tensor_matrix.rst [hunk 11]
```diff
@@ -24,7 +24,7 @@ Tensors of scalars
 
 * Tensor values are initially zero.
 * Sparse tensors are initially inactive.
-* See :ref:`scalar_tensor` for more details.
+* See :ref:`scalar_field` for more details.
```
Comment (archibate) [inline docs/tensor_matrix.rst:27]:
```suggestion
* See :ref:`scalar_tensor` for more details.
```
Was simply run `sed s/ti.var/ti.field/g` so hard?

**Comment by Rullec** [comment]
Maybe I misunderstood your meaning?
I convert all of `ti.var` into `ti.field`, as well as **some related texts near the code snippet**. It is what I understood and also what I wanted to do.
But it seems your meaning is **ONLY** do `ti.var` -> `ti.field`?

**Comment by Rullec** [comment]
Now I overwrote these stuff and simply did replacement from `ti.var` to `ti.field`, with nothing changed anymore. 

Maybe we need a quite clear plan for this documents modification...

**Comment by archibate** [comment]
> But it seems your meaning is ONLY do ti.var -> ti.field?

Yes, I intentionally make this PR incomplete.
Did you noticed the PR destination?
It's not master, it's rullec-field.
So it's okay to merge incomplete confusing changes for easy-reviewing.
Once all sub PRs are merged into rullec-field, we can merge this complete but large changeset into master without review.

**Comment by Rullec** [comment]
Very nice and thanks for your explanation! It confused me indeed.
It seems now all `ti.var` has been replaced by `ti.field`, the task of this PR is finished. What can I or When can I move forward to have next substep?

**Comment by archibate** [comment]
> What can I

Now please run:
```bash
git checkout master
git pull
git checkout -b rullec-field
git pull --set-upstream https://github.com/taichi-dev/taichi.git rullec-field
git checkout -b new_allocating_api2
```
and begin your work.

**Comment by Rullec** [comment]
Thanks!

彭于斌 <notifications@github.com> 于2020年7月18日周六 下午9:00写道：

**Pull request merged**